src/main/java/com/amplitude/Retry.java

```diff
@@ -0,0 +1,254 @@
+      if (onceReponse.RateLimitBody != null) {
+        JSONObject exceededDailyQuotaUsers =
+            onceReponse.RateLimitBody.getJSONObject("exceededDailyQuotaUsers");
+        JSONObject exceededDailyQuotaDevices =
+            onceReponse.RateLimitBody.getJSONObject("exceededDailyQuotaDevices");
+        if ((userId.length() > 0 && exceededDailyQuotaUsers.has(userId))
+            || (deviceId.length() > 0 && exceededDailyQuotaDevices.has(deviceId))) {
+          shouldRetry = false;
+        }
+      }
+      // Reduce the payload to reduce risk of throttling
+      shouldReduceEventCount = true;
+    } else if (onceReponse.status == Status.PAYLOAD_TOO_LARGE) {
+      shouldRetry = true;
+    } else if (onceReponse.status == Status.INVALID) {
+      if (events.size() == 1) {
+        shouldRetry = false;
+      } else {
+        eventIndicesToRemove = collectInvalidEventIndices(onceReponse);
+      }
+    } else if (onceReponse.status == Status.SUCCESS) {
+      shouldRetry = false;
+    }
+    return new RetryEventsOnceResult(shouldRetry, shouldReduceEventCount, eventIndicesToRemove);
+  }
+
+  private static void retryEventsOnLoop(String userId, String deviceId, String apiKey) {
+    List<Event> eventsBuffer =
+        (idToBuffer.get(userId) != null) ? idToBuffer.get(userId).remove(deviceId) : null;
+    if (eventsBuffer == null || eventsBuffer.size() == 0) {
+      cleanUpBuffer(userId);
+      return;
+    }
+    int retryTimes = Constants.RETRY_TIMEOUTS.length;
+    Thread retryThread =
+        new Thread(
+            () -> {
+              int eventCount = eventsBuffer.size();
+              for (int numRetries = 0; numRetries < retryTimes; numRetries++) {
+                long sleepDuration = Constants.RETRY_TIMEOUTS[numRetries];
+                try {
+                  Thread.sleep(sleepDuration);
+                  boolean isLastTry = numRetries == Constants.RETRY_TIMEOUTS.length - 1;
+                  List<Event> eventsToRetry = eventsBuffer.subList(0, eventCount);
+                  RetryEventsOnceResult retryResult =
+                      retryEventsOnce(userId, deviceId, eventsToRetry, apiKey);
+                  boolean shouldRetry = retryResult.shouldRetry;
+                  boolean shouldReduceEventCount = retryResult.shouldReduceEventCount;
+                  int[] eventIndicesToRemove = retryResult.eventIndicesToRemove;
+                  if (eventIndicesToRemove.length > 0) {
+                    int numEventsRemoved = 0;
+                    for (int i = 0; i < eventIndicesToRemove.length; i++) {
+                      int index = eventIndicesToRemove[i];
+                      if (index < eventCount) {
+                        eventsBuffer.remove(i);
+                        numEventsRemoved += 1;
+                      }
+                    }
+                    eventCount -= numEventsRemoved;
+                    eventsInRetry.set(eventsInRetry.intValue() - numEventsRemoved);
```

Comment:
when we update eventsInRetry, could we use method like `decrementAndGet` to guarantee the thread safety? Two atomic operation together won't guarantee the atomic though.

Reply:
decrementAndGet only decrements by one. But I didn't find a way to decrease a certain number except using .set() method.

Reply:
I see, could we use addAndGet with negative value? or maybe updateAndGet?